### ISSUE TITLE
Replace `clang` with ANTLR in `parser`

### ISSUE THREAD

**Issue by nuthanmunaiah** [samaritan/services#18]
# Description

The `parser` service currently uses `clang` to parse C/C++ file to enumerate the functions defined in it. However, since all that the `parser` service needs is to _parse_ a C/C++ file, using `clang` seems like an overkill. Furthermore, extending the `parser` service to add the ability to parse files written in languages other than C/C++ will require a parser like `clang` for the language in question.

An alternative to `clang` is to use [ANTLR](https://www.antlr.org/) to generate a parser for all languages that we want to support in the `parser` service. The use of ANTLR also affords us the ability to extend the `parser` service using the extensive set of [grammars](https://github.com/antlr/grammars-v4) available for ANTLR.

**Comment by nuthanmunaiah** [comment]
**TL;DR** Use [SrcML](https://www.srcml.org/) instead of ANTLR to replace `clang` as the source code parser in the `parser` service.

`parser` service was refactored to use ANTLR instead of `clang` to enumerate functions declared and/or defined in C/C++ source code. We, however, observed considerable differences when functions identified by `clang` and that identified by ANTLR were compared. In the comparison, each function was uniquely identified as a string that followed the pattern `<name>@<begin>:<end>`, where `<name>` is the name of the function and `<begin>` and `<end>` are the line numbers at which the declaration or definition of the function begins and ends, respectively. We used [FFmpeg at 20fed2f](https://github.com/FFmpeg/FFmpeg/tree/20fed2f0ab197d60801280dfc844f6b29a397ff2) as the subject in the comparison. 

Shown in the tables below are summary of the number of functions identified per file at different levels. The **# Functions Missed by ANTLR** is of particular interest to demonstrate the severity of the differences between `clang` and ANTLR.

We chose to use [SrcML](https://www.srcml.org/) as the source code parser in the `parser` service to overcome the limitations of ANTLR.

## # Functions Identified by `clang`

|  Minimum  |    Mean    |   Median   |  Maximum  |    Total    |  Variance  |
| --------: | ---------: | ---------: | --------: | ----------: | ---------: |
|      0.00 |      10.65 |       5.00 |  1,299.00 |   34,281.00 |     991.21 |

## # Functions Identified by ANTLR

|  Minimum  |    Mean    |   Median   |  Maximum  |    Total    |  Variance  |
| --------: | ---------: | ---------: | --------: | ----------: | ---------: |
|      0.00 |       3.24 |       1.00 |     88.00 |   10,445.00 |      35.99 |

## # Functions Identified by both `clang` and ANTLR

|  Minimum  |    Mean    |   Median   |  Maximum  |    Total    |  Variance  |
| --------: | ---------: | ---------: | --------: | ----------: | ---------: |
|      0.00 |       1.79 |       0.00 |     75.00 |    5,749.00 |      18.87 |

## # Functions Missed by ANTLR

|  Minimum  |    Mean    |   Median   |  Maximum  |    Total    |  Variance  |
| --------: | ---------: | ---------: | --------: | ----------: | ---------: |
|      0.00 |       8.86 |       3.00 |  1,299.00 |   28,532.00 |     972.86 |
